### PR TITLE
Fixes issue where workflow execution hangs if multiple parallel local activities are spawned

### PR DESCRIPTION
**Pull request by mastermanu**
WARNING: This PR breaks any existing Workflows that have a Local Activity or Side Effect in its history.

- Fixes (https://github.com/temporalio/go-sdk/issues/228) by using a separate ID counters for local activities and side effects. 

- Adds a few more integration tests w/replay so that we can catch any potential backwards incompatible changes in the future.

### CODE REVIEW

File: internal/internal_decision_state_machine.go [hunk 1]
```diff
@@ -755,7 +755,7 @@ func (h *commandsHelper) addCommand(command commandStateMachine) {
 	element := h.orderedCommands.PushBack(command)
 	h.commands[command.getID()] = element
 
-	// Every time new command is added increment the counter used for generating ID
+	// Every time new command is added increment t6he counter used for generating ID
```
Comment (mastermanu) [inline internal/internal_decision_state_machine.go:758]: oops

**Comment by mastermanu** [comment]
@samarabbas to run the integration test locally and see if there is an obvious, simple, safe solution

File: test/workflow_test.go [hunk 2]
```diff
@@ -654,6 +655,102 @@ func (w *Workflows) WorkflowWithLocalActivityCtxPropagation(ctx workflow.Context
...
+		futures = append(futures, workflow.ExecuteLocalActivity(ctx, activities.Echo, i))
+	}
+
+	for i, future := range futures {
+		var activityResult int
+		if err := future.Get(ctx, &activityResult); err != nil {
+			return "", err
+		}
+
+		if activityResult != i {
```
[the remark this answers is not in the document]
Comment (mastermanu) [reply test/workflow_test.go:673]: the activity is echoing the input given to it. the order of the futures in the array will be fixed, so this condition is correct

File: test/workflow_test.go [hunk 3]
```diff
@@ -654,6 +655,102 @@ func (w *Workflows) WorkflowWithLocalActivityCtxPropagation(ctx workflow.Context
...
+		})
+	}
+
+	for i, future := range futures {
+		var sideEffectValue int
+		if err := future.Get(ctx, &sideEffectValue); err != nil {
+			return "", err
+		}
+
+		if i != sideEffectValue {
```
[the remark this answers is not in the document]
Comment (mastermanu) [reply test/workflow_test.go:730]: the test doesn't make any assumptions about the order of goroutine executions given that the futures will always be ordered by 'i'

File: internal/internal_event_handlers.go [hunk 4]
```diff
@@ -110,6 +110,13 @@ type (
 		unstartedLaTasks  map[string]struct{}
 		openSessions      map[string]*SessionInfo
 
+		// LocalActivities have a separate, individual counter instead of relying on actual commandEventIDs.
+		// This is because command IDs are only incremented on activity completion, which breaks
+		// local activities that are spawned in parallel as they would all share the same command ID
+		localActivityCounterID int64
+
+		sideEffectCounterID int64
```
[the remark this answers is not in the document]
Comment (mastermanu) [reply internal/internal_event_handlers.go:118]: i thought for debugging / cleanliness it was much easier to do this. Now you'll be able to look at markers and tell that this was the "2nd local activity" vs the "3rd side effect." If we feel strongly about this though, can merge it back into a single id counter (which is what I initially had)

File: test/integration_test.go [hunk 5]
```diff
@@ -475,6 +475,38 @@ func (ts *IntegrationTestSuite) TestWorkflowWithLocalActivityCtxPropagation() {
...
+func (ts *IntegrationTestSuite) TestWorkflowWithParallelLocalActivitiesUsingReplay() {
+	replayer := worker.NewWorkflowReplayer()
+	replayer.RegisterWorkflowWithOptions(ts.workflows.WorkflowWithParallelLocalActivities, workflow.RegisterOptions{DisableAlreadyRegisteredCheck: true})
+	err := replayer.ReplayWorkflowHistoryFromJSONFile(ilog.NewDefaultLogger(), "replaytests/parallel-local-activities.json")
+	ts.NoError(err)
+}
+
+func (ts *IntegrationTestSuite) TestWorkflowWithLongLocalActivityAndHeartbeat() {
+	if !ts.config.IsStickyOff {
+		ts.NoError(ts.executeWorkflow(uuid.New(), ts.workflows.WorkflowWithLongLocalActivityAndHeartbeat, nil))
```
Comment (mastermanu) [inline test/integration_test.go:491]: let's use proper name here

File: internal/internal_event_handlers.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply internal/internal_event_handlers.go:118]: I don't feel strongly

File: test/workflow_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mfateev) [reply test/workflow_test.go:673]: Right. I missed that the features are by order of creation and not completion

**Pull request merged**